STM32F105xx Support

attie-argentum opened the pull request:
I've been working on support for STM32F105xx series parts. Please consider this a WIP... I'll be updating it as I go, but I'd like feedback on what I consider some fairly critical / core changes that I've made.

My todo list:

- [x] Clock Tree
- [x] USB Device
- [x] CAN1 Support
- [ ] ~~CAN2 Support (probably not a small task - shared mailboxes, etc...)~~ - I'll look into this later
- [ ] ~~ADC Support (I expect the clock configuration to need work)~~ - I'll look into this later

I'm not sure about some of my changes, and would appreciate input from others... mainly:

- a3df0d4 - this feels very wrong to me. I couldn't get the `SystemCoreClock` variable to get a sensible value via the Zephyr code, however using `SystemCoreClockUpdate()` sets it correctly.
- 0d504b2 - I'm not 100% about some of my changes in `clock_stm32_ll_common.c` and their suitability for other parts... Please review it carefully and test this on other parts.

Unfortunately the board I'm targeting isn't open, so I can't expect others to have the same platform as myself... It should be largely irrelevant though.

attie-argentum committed:
dts: arm: stm32: add support for STM32F105xx parts

The STM32F105xx is effectively an STM32F107xx without Ethernet. As such,
it's possible to include the STM32F105's device tree from STM32F107, and
add in any additional nodes (currently just DMA2, though according to
the datasheet, this may be supported by the STM32F105xx and other parts
too).

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
usb: stm32: added support for USB Device mode on STM32F105xx parts

The STM32F105xx USB clock goes through a prescaler either PLL1 x2 /2
or PLL1 x2 /3, the output of this must be 48 MHz. As such, the output
of PLL1 must be 48 MHz or 72 MHz.

NOTE: This requires that the system is running from PLL1 (PLLCLK).
      Support for running from PLL2 will be implemented in the future.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
dts: arm: stm32: added support for USB Device mode on STM32F105xx parts

Added device tree nodes for USB Device mode support on STM32F105xx parts

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
dts: stm32: added CAN1 & CAN2 support for STM32F105xx parts

Added device tree definition for CAN 1 and CAN 2 in STM32F105

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

erwango reviewed (changes requested):
@attie-argentum , thanks for this contribution. This looks generally good.
I had a very quick review, please check the comments. 
I'll check more in detail for the `SystemCoreClockUpdate` thing.
I'll be interested to know in which cases you're seeing failures (we'd like to get rid of `SystemCoreClock` in general).

Besides, please have a check on contribution guidelines for the commit messages styles and fix them to get zephyr-bot happy.

PS: I haven't checked the USB and CAN stuff yet.

attie-argentum commented on dts/arm/st/f1/stm32f105.dtsi:
```diff
@@ -0,0 +1,77 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <st/f1/stm32f1.dtsi>
```
It is, yes, though it doesn't look like the STM32F107xx has Ethernet support in Zephyr at the moment...
- [dts/arm/st/f1/stm32f107.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/st/f1/stm32f107.dtsi)
- [dts/arm/st/f1/stm32f107Xc.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/st/f1/stm32f107Xc.dtsi)

erwango replied:
Ok.
On stm32 family, we try to take advantage of SoC commonalities to optimize size of code and ease maintenance of dtsi files. 
According to that model, in order to add STM32F105 description, we should have stm32f107.dtsi renamed as stm32f105.dtsi and new stm32f107.dtsi file that includes stm32f105.dtsi. And once ethernet will be supported on stm32f1 series we'll add eth node in stm32f107 node (and until that happen, it will remain empty).

attie-argentum commented:
Thanks for the comments @erwango 

> I'll check more in detail for the SystemCoreClockUpdate thing. I'll be interested to know in which cases you're seeing failures (we'd like to get rid of SystemCoreClock in general).

I don't have figures to hand (I'll get them soon - JTAG is required), but IIRC the `SystemCoreClock` is set to something _very_ low, in the hundreds of Hz. USB, Serial (and thus `prink()`) and CAN all fail to come up when I revert this change. I have a feeling it's related to PLL2 in the clock tree, which isn't being accounted for... but I remember the value being surprisingly low, so it may be more than this.

> Besides, please have a check on contribution guidelines for the commit messages styles and fix them to get zephyr-bot happy.

I've been through a few rebases, hopefully it's more-or-less in line now... I'll review the guidelines too though.

jfischer-phytec-iot commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -254,6 +254,25 @@ static int usb_dc_stm32_clock_enable(void)
 		LOG_ERR("Unable to set USB clock source to PLL.");
 	}
 #endif /* CONFIG_CLOCK_STM32_MSI_PLL_MODE && !CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI */
+
```
Changes in usb_dc_stm32.c should be in specific commit.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -361,6 +364,36 @@ static int stm32_clock_control_init(struct device *dev)
+	/* switch the input to use PLL2 instead of HSE directly */
+	s_PLLInitStruct.Prediv |= RCC_CFGR2_PREDIV1SRC << 4U;
+
+	/* Configure and enable PLL2 */
```
Configure but not enable.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -361,6 +364,36 @@ static int stm32_clock_control_init(struct device *dev)
+		* CONFIG_CLOCK_STM32_PLL2_MULTIPLIER;
+
+	/* switch the input to use PLL2 instead of HSE directly */
+	s_PLLInitStruct.Prediv |= RCC_CFGR2_PREDIV1SRC << 4U;
```
I think this could be done directly in `config_pll2_init`.

attie-argentum replied:
I didn't want to contaminate things too much... it should probably technically live in `config_pll_init()`, though shouldn't be set when the PLL is clocked by HSE (or anything other than PLL2).

The `RCC_CFGR2_PREDIV1SRC << 4U` is a bit _magic_... but comes from here:
- [stm32f1xx_ll_rcc.h:426](https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L426)
- [stm32f1xx_ll_rcc.h:443](https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L443)

Perhaps worth ORing [`LL_RCC_PLLSOURCE_PLL2`](https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L422) over `pllint->Prediv` in [`config_pll_init()`](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/clock_control/clock_stm32f1.c#L33) if `CONFIG_CLOCK_STM32_PLL_SRC_PLL2` is defined?

erwango replied:
I think it is ok to put it in `config_pll2_init` since dedicated to pll2, this shouldn't have much impact

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -254,6 +254,25 @@ static int usb_dc_stm32_clock_enable(void)
+	 *
+	 * this presumes that the system is running from PLLCLK
+	 */
+	if (SystemCoreClock == 48000000U) {
```
As mentioned, we're trying to get rid of `SystemCoreClock` variable which is available only on ARM platform.
Please use `sys_clock_hw_cycles_per_sec`.

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -254,6 +254,25 @@ static int usb_dc_stm32_clock_enable(void)
 	}
 #endif /* CONFIG_CLOCK_STM32_MSI_PLL_MODE && !CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI */
+
+#elif defined(RCC_CFGR_OTGFSPRE) \
```
Please add a comment clarifying the is F105/F107 specific code

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -254,6 +254,25 @@ static int usb_dc_stm32_clock_enable(void)
 #endif /* CONFIG_CLOCK_STM32_MSI_PLL_MODE && !CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI */
+
+#elif defined(RCC_CFGR_OTGFSPRE) \
+	&& defined(LL_RCC_USB_CLKSOURCE_PLL_DIV_2) \
```
`LL_RCC_USB_CLKSOURCE_PLL_DIV_2` and `LL_RCC_USB_CLKSOURCE_PLL_DIV_3` defined under `RCC_CFGR_OTGFSPRE`.
So first line should be enough

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -254,6 +254,25 @@ static int usb_dc_stm32_clock_enable(void)
+	 *   - PLLCLK * 2 / 2
+	 *   - PLLCLK * 2 / 3
+	 *
+	 * this presumes that the system is running from PLLCLK
```
Please add a check to verify it and error out if not the case.

erwango reviewed (changes requested):
Few additional comments.
Also please refactor your commits so we're not merging code and its fix but only correct code (at least on par with the current knowledge).

Last, you still have few issues on the commit messages:
- avoid multilines titles (titles with"...") , even not stated in the guidelines
- Titles should begin with subsystem (avoid "FIXUP: ..")

ydamigos commented on soc/arm/st_stm32/stm32f1/dts_fixup.h:
```diff
@@ -153,13 +153,23 @@
+#define DT_USB_RAM_SIZE			DT_ST_STM32_OTGFS_50000000_RAM_SIZE
+#define DT_USB_CLOCK_BITS		DT_ST_STM32_OTGFS_50000000_CLOCK_BITS
+#define DT_USB_CLOCK_BUS		DT_ST_STM32_OTGFS_50000000_CLOCK_BUS
+#elif USB_OTG_HS
```
It should be `#elif USB`. These macros are defined for USB 2.0 FS device interface (STM32F102/STM32F103). `USB_OTG_HS` is for SoCs with OTG high speed controller.

attie-argentum replied:
The STM32F102 and F103 have the USB full-speed peripheral (i.e: Chapter 23 of the datasheet)
The STM32F105 and F107 have the USB OTG_FS peripheral (i.e: Chapter 28 of the datasheet)

I think this is correct?... apologies if not.

ydamigos replied:
> I think this is correct?

Yes. There is no USB OTG_HS peripheral in STM32F1xx series.

ydamigos commented on soc/arm/st_stm32/stm32f1/dts_fixup.h:
```diff
@@ -209,6 +219,21 @@
+#define DT_CAN_1_PHASE_SEG1		DT_ST_STM32_CAN_40006400_PHASE_SEG1
+#define DT_CAN_1_PHASE_SEG2		DT_ST_STM32_CAN_40006400_PHASE_SEG2
+#define DT_CAN_1_CLOCK_BUS		DT_ST_STM32_CAN_40006400_CLOCK_BUS
+#define DT_CAN_1_CLOCK_BITS		DT_ST_STM32_CAN_40006400_CLOCK_BITS
```
Why do we define them again?

attie-argentum replied:
Ah, my work was before c180e05, thanks for pointing that out

attie-argentum commented:
> Also please refactor your commits so we're not merging code and its fix but only correct code (at least on par with the current knowledge).

> Titles should begin with subsystem (avoid "FIXUP: ..")

Apologies, the `FIXUP: ...` commit is "_to be fixed_" rather than a fix to one of my previous changes - it's around the use of `SystemCoreClock`.

attie-argentum commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -332,11 +351,13 @@ static int usb_dc_stm32_init(void)
 #endif
 	usb_dc_stm32_state.pcd.Init.dev_endpoints = DT_USB_NUM_BIDIR_ENDPOINTS;
 	usb_dc_stm32_state.pcd.Init.speed = usb_dc_stm32_get_maximum_speed();
+#ifndef USB_OTG_FS
```
Because [`USB_OTG_CfgTypeDef`](https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_usb.h#L105) (for STM32F1xx with `USB_OTG_FS` defined) don't have a `phy_itface` member... (they do when `USB` is defined)

I might have got something wrong here...

attie-argentum replied:
For reference:
```
/home/user/zephyrproject/zephyr/drivers/usb/device/usb_dc_stm32.c: In function 'usb_dc_stm32_init':
/home/user/zephyrproject/zephyr/drivers/usb/device/usb_dc_stm32.c:357:29: error: 'PCD_InitTypeDef {aka struct <anonymous>}' has no member named 'phy_itface'
  usb_dc_stm32_state.pcd.Init.phy_itface = PCD_PHY_EMBEDDED;
                             ^
```

ydamigos replied:
You are right. I totally missed that. However, it may break the driver for other series (e.g. L4) where `USB_OTG_CfgTypeDef` has a `phy_itface` member.

attie-argentum replied:
Eeh, I was concerned about that. I'll look at options and get back to you :+1:

attie-argentum replied:
Variants of the `USB_OTG_CfgTypeDef` structure:
- `.../stm32f1xx/drivers/include/stm32f1xx_ll_usb.h` - `phy_itface` *not present*
- `.../stm32f2xx/drivers/include/stm32f2xx_ll_usb.h` - `phy_itface` present
- `.../stm32f4xx/drivers/include/stm32f4xx_ll_usb.h` - `phy_itface` present
- `.../stm32f7xx/drivers/include/stm32f7xx_ll_usb.h` - `phy_itface` present
- `.../stm32h7xx/drivers/include/stm32h7xx_ll_usb.h` - `phy_itface` present
- `.../stm32l4xx/drivers/include/stm32l4xx_ll_usb.h` - `phy_itface` present

This makes STM32F1xx the outlier...

I think the correct solution here is to swap `#ifndef USB_OTG_FS` for `#ifndef CONFIG_SOC_SERIES_STM32F1X` (as per `soc/arm/st_stm32/stm32f1/Kconfig.series`).

Let me know what you think.

ydamigos replied:
It is include after the latest update of stm32 modules. Did you update?

erwango commented on soc/arm/st_stm32/stm32f1/Kconfig.defconfig.stm32f105xc:
```diff
@@ -0,0 +1,25 @@
+# Kconfig - ST Microelectronics STM32F105XC MCU
+#
```
Remove blank lines in Kconfig headers (recent common request).

attie-argentum replied:
Copied the header format from `Kconfig.defconfig.stm32f107xc`

erwango commented on dts/arm/st/f1/stm32f105Xc.dtsi:
```diff
@@ -30,6 +30,23 @@
 			status = "disabled";
 			label = "OTGFS";
 		};
+
```
Update commit messge

attie-argentum replied:
Please could you expand on what you mean by this?

erwango reviewed (changes requested):
@attie-argentum do you intend to complete the work? If yes, a rebase would be required + feedbacks on pending comments. Thanks

attie-argentum commented:
Hi @erwango - yes, very sorry for the delays on my part. I will be looking at this today, hoping to resolve all comments that I can. I'll rebase too.

attie-argentum commented:
Hi all,

I believe I've resolve all of the requested changes / queries.
Please let me know if you have any further comments.

ydamigos commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -318,11 +345,13 @@ static int usb_dc_stm32_init(void)
 #else
 	usb_dc_stm32_state.pcd.Init.phy_itface = PCD_PHY_EMBEDDED;
 #endif /* DT_COMPAT_ST_STM32_USBPHYC */
+#endif /* !USB_OTG_FS */
```
!CONFIG_SOC_SERIES_STM32F1X

attie-argentum replied:
🤦‍♀️

ydamigos commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -318,11 +345,13 @@ static int usb_dc_stm32_init(void)
 #endif
 	usb_dc_stm32_state.pcd.Init.dev_endpoints = DT_USB_NUM_BIDIR_ENDPOINTS;
 	usb_dc_stm32_state.pcd.Init.speed = usb_dc_stm32_get_maximum_speed();
+#ifndef CONFIG_SOC_SERIES_STM32F1X
```
Did you check with the latest STM32Cube modules? It seems they added `phy_itface` to `USB_OTG_CfgTypeDef` structure.

attie-argentum replied:
Ah yes, I missed the update, thanks!

attie-argentum commented:
... please hold off on this - I've been having issued with getting stuck in `uart_poll_out()`, and now believe my patches to be part of the problem(?), though I'm not currently sure why.

```
^C
Program received signal SIGTRAP, Trace/breakpoint trap.
0x080069f2 in uart_stm32_poll_out (dev=0x20003308 <__device_uart_stm32_USART_1>, c=42 '*') at /home/user/zephyrproject/zephyr/drivers/serial/uart_stm32.c:400
400             while (!LL_USART_IsActiveFlag_TXE(UartInstance)) {
(gdb) bt
#0  uart_stm32_poll_out (dev=0x20003318 <__device_uart_stm32_USART_1>, c=42 '*') at /home/user/zephyrproject/zephyr/drivers/serial/uart_stm32.c:400
#1  0x080016b6 in z_impl_uart_poll_out (out_char=42 '*', dev=<optimized out>) at /home/user/zephyrproject/zephyr/include/drivers/uart.h:644
#2  uart_poll_out (out_char=42 '*', dev=<optimized out>) at zephyr/include/generated/syscalls/uart.h:112
#3  console_out (c=42) at /home/user/zephyrproject/zephyr/drivers/console/uart_console.c:91
#4  0x080051c2 in z_vprintk (out=out@entry=0x8000ae5 <char_out>, ctx=ctx@entry=0x2000227c <z_main_stack+964>, fmt=0x8007b6f "** Booting Zephyr OS build %s %s ***\n",
    fmt@entry=0x8007b6e "*** Booting Zephyr OS build %s %s ***\n", ap=...) at /home/user/zephyrproject/zephyr/lib/os/printk.c:118
#5  0x08000bae in vprintk (fmt=fmt@entry=0x8007b6e "*** Booting Zephyr OS build %s %s ***\n", ap=..., ap@entry=...) at /home/user/zephyrproject/zephyr/lib/os/printk.c:352
#6  0x08005476 in printk (fmt=0x8007b6e "*** Booting Zephyr OS build %s %s ***\n") at /home/user/zephyrproject/zephyr/lib/os/printk.c:405
#7  0x08003b50 in bg_thread_main (unused1=<optimized out>, unused2=<optimized out>, unused3=<optimized out>) at /home/user/zephyrproject/zephyr/kernel/init.c:252
#8  0x0800548c in z_thread_entry (entry=0x8003b39 <bg_thread_main>, p1=<optimized out>, p2=<optimized out>, p3=<optimized out>) at /home/user/zephyrproject/zephyr/lib/os/thread_entry.c:29
#9  0x08002710 in arch_switch_to_main_thread (main_thread=0x8007b6e, main_stack=<optimized out>, main_stack_size=<optimized out>, _main=0x0) at /home/user/zephyrproject/zephyr/arch/arm/core/aarch32/thread.c:428
```

attie-argentum commented:
Turns out it's really quite important that we call `SystemCoreClockUpdate()` in `stm32_clock_control_init()` when we're running from PLL2, as per my "_FIXUP:_" commit (a3df0d4). I've adjusted my patches so that the USB code doesn't require it any more, but the UART driver does (and anything that calls on `clock_control_get_rate()` will do too).

Without it, and while running from PLL2, the UART's baudrate gets set to `0x0`, the line idles low, and no frames are transmitted.

Without the call to `SystemCoreClockUpdate()`, we get a value of `288` back from `clock_control_get_rate()` (where it should be `48000000`)... This is the "_very low, in the hundreds of Hz_" value that I mentioned previously.

`clock_control_get_rate()` calls through to `stm32_clock_control_get_subsys_rate()`, which in turn uses the  forbidden `SystemCoreClock` ([clock_stm32_ll_common.c:194](https://github.com/zephyrproject-rtos/zephyr/blob/9e92bab/drivers/clock_control/clock_stm32_ll_common.c#L194))

I don't think it's appropriate to address this in this patchset, so would calling `SystemCoreClockUpdate()` when running from PLL2 be acceptable with appropriate commentary?

erwango commented:
CI failed due to: https://github.com/zephyrproject-rtos/zephyr/issues/21568

attie-argentum commented:
Thanks @erwango ... is there any way for me to rerun it? Are you able to?

erwango commented:
> Thanks @erwango ... is there any way for me to rerun it? Are you able to?

"recheck" used to work, but seems it doesn't anymore since few months.
What you can do is rebase and push again.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -368,6 +371,36 @@ static int stm32_clock_control_init(struct device *dev)
+		PLL2Frequency,
+		hse_bypass,
+		&s_PLLInitStruct,
+		&s_ClkInitStruct);
```
Don't we miss APB/AHB configuration at this point ?

attie-argentum replied:
[`UTILS_EnablePLLAndSwitchSystem()`](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L517) seems to set [AHB](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L564), [APB1](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L572) and [APB2](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L573) for us, and it is called from inside `LL_PLL_ConfigSystemClock_HSE()` (i.e: here), so I think we're okay...

Certainly the PLL code from before this patch didn't deal with APB / AHB config.

Only the non-PLL sources seem to set the prescalers themselves for some reason... (i.e: `CONFIG_CLOCK_STM32_SYSCLK_SRC_HSE` / `CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI` / `CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI`)

erwango replied:
Right, I got confused with non PLL feed sysclock configs.
This is fine, my bad.

erwango replied:
I'm reopening the point as you can't use `LL_PLL_ConfigSystemClock_HSE()` in your case.
So you don't have access to `UTILS_EnablePLLAndSwitchSystem()`

erwango reviewed (changes requested):
Looks good in general.
Some remaining concerns on PLL configuration with PLL2 as source.

attie-argentum commented:
While looking in to @erwango 's APB/AHB comment, I did discover the source of the `SystemCoreClock` issues...

Given that I've calculated `PLL2Frequency` [here](https://github.com/zephyrproject-rtos/zephyr/pull/21224/files#diff-f28d8c15aadb6858ad62ec32e4c7f61fR386), I expected that the resulting `pllfreq` in `LL_PLL_ConfigSystemClock_HSE()` would have been correct, but it isn't... it's `228`(!)

```
pllfreq = __LL_RCC_CALC_PLLCLK_FREQ(PLL_InputFrequency / (UTILS_PLLInitStruct->Prediv + 1U), UTILS_PLLInitStruct->PLLMul);
        = __LL_RCC_CALC_PLLCLK_FREQ(     40000000      / (          1048580           + 1U),           1048576          );
        = ( ( 40000000 / ( 1048580 + 1U) ) * ( ( (  1048576 & RCC_CFGR_PLLMULL ) >> RCC_CFGR_PLLMULL_Pos ) + 2U );
        = ( ( 40000000 / ( 1048580 + 1U) ) * ( ( ( 0x100000 &     0x3C0000     ) >>          18          ) + 2U );
        = ( ( 40000000 / ( 1048580 + 1U) ) * ( ( (         0x100000            ) >>          18          ) + 2U );
        = ( ( 40000000 / ( 1048580 + 1U) ) * ( (                                4                        ) + 2U );
        = ( 40000000 / 1048581 ) * ( 6 );
        = ( 38 ) * ( 6 );
        = 228;
```

The issue here is that we OR in `RCC_CFGR2_PREDIV1SRC << 4U` (`1 << 16` / `0x10000`) to make the system run from PLL2 as the source, as per [`LL_RCC_PLL_ConfigDomain_SYS()`](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L1438)... this mucks up the calculations in [`UTILS_GetPLLOutputFrequency()`](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L453).

I don't see a good way to handle this, without patching the STM32Cube files - which I'd imagine you don't want to do?

attie-argentum commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -491,6 +524,14 @@ static int stm32_clock_control_init(struct device *dev)
+	 * incorrect clock rate can be provided (i.e: 288 Hz instead of 48 MHz). */
+	SystemCoreClockUpdate();
+#endif /* CONFIG_CLOCK_STM32_PLL_SRC_PLL2 */
+
```
I forgot to mention my look into using your alternative to `SystemCoreClockUpdate()`... I think this won't work (see [my comment)](https://github.com/zephyrproject-rtos/zephyr/pull/21224#issuecomment-578225889)... I've moved it, but are you content to stick with `SystemCoreClockUpdate()` for now?

erwango commented:
> Given that I've calculated PLL2Frequency here, I expected that the resulting pllfreq in LL_PLL_ConfigSystemClock_HSE() would have been correct, but it isn't... it's 228(!)

@attie-argentum, `LL_PLL_ConfigSystemClock_HSE()` should not be used in your case. It is meant to be used with HSE as direct PLL source ("This function configures system clock with HSE as clock source of the PLL"). And unfortunately, there is no equivalent function for PLL2 as source of PLL.
So you need to  do the equivalent actions on your own.

attie-argentum commented:
> `LL_PLL_ConfigSystemClock_HSE()` should not be used in your case. It is meant to be used with HSE as direct PLL source.

Hmm... [`LL_PLL_ConfigSystemClock_HSE()`](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L331) [calls](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c#L371) [`LL_RCC_PLL_ConfigDomain_SYS()`](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L1477), which seems to be happy to take an argument of [`LL_RCC_PLLSOURCE_PLL2_DIV_5`](https://github.com/zephyrproject-rtos/hal_stm32/blob/fc5cb1a/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L431) (i.e: use PLL2's output, div 5).

Is it not okay to make use of this?

erwango commented:
@attie-argentum, 

> Is it not okay to make use of this? (just pointing it out in case you missed it...)

Indeed. This is what I pointed to in https://github.com/zephyrproject-rtos/zephyr/pull/21224#issuecomment-578719188

erwango commented:
@attie-argentum any plans for this PR ?

attie-argentum commented:
Hi @erwango !

This is still on my list, Unfortunately I've been very the last few(!) months...
I'll do what I can to address your outstanding comment and update in the next week or two.

erwango commented:
@attie-argentum, no rush, we're in code freeze right now. So we'll have to wait V2.3.0 release before integrating this change.

attie-argentum commented:
Hi @erwango 

I'm (finally) looking into addressing your comments about clock configuration and avoiding the use of `LL_PLL_ConfigSystemClock_HSE()`.

I'm a bit uneasy about how much I'm having to bring into zephyr from the `hal_stm32` codebase... static functions, and macros in C files, etc...

- I expect that we want to avoid putting any changes in `hal_stm32` that didn't originate from ST - is that correct?
- Are you okay for me to be adding (duplicating) what feels like quite a lot into `clock_stm32_ll_common.c` for this purpose?
- Is there a better approach that I'm not seeing?

Attie

attie-argentum commented:
I consider both [a6910d5](https://github.com/zephyrproject-rtos/zephyr/pull/21224/commits/a6910d57170301c783204bda162706ccf11f25b3) and [d45719f](https://github.com/zephyrproject-rtos/zephyr/pull/21224/commits/d45719fbfb2ec1ccb98f15b2a8759c9a4e5071f9) to be WIP... they'll likely need to be merged into one, but I wanted to be clear about what's required to get this to work:

1. Quite a fair chunk of the HAL STM32 codebase has been duplicated into the kernel.
2. A fix is required in HAL STM32 to properly support calculations when configured to run from PLL2 (rather than HSE directly)

attie-argentum commented:
@erwango / @jfischer-phytec-iot / @ydamigos - any comments on my latest updates?

attie-argentum commented:
@erwango / @jfischer-phytec-iot / @ydamigos - Hi all, sorry to ping you again... I was wondering if there are any comments on the notes I shared above?

erwango commented:
@attie-argentum Sorry I missed your previous ping. At this point, I think it would be good to get back to the starting point:

IIRC:
- for some reason, you need PLL to be driven from PLL2
- atm, current LL code doesn't allow proper PLL2 calculation. This is due to both missing and erroneous code 
Am I correct ?

Before going further, if it is clear there are LL issues, I'd propose to first have a try to fix this (or at least have initial feedback) at the source https://github.com/STMicroelectronics/STM32CubeF1  (I'll  try to ensure you get a response).

erwango commented:
@attie-argentum Btw, I think there are non controversial parts in this PR that would have no issue to go upstream (dts, soc Kconfig). Feel free to push a PR with these

attie-argentum commented:
@erwango - no problem, thanks for your response!

Your summary is correct.

> I'd propose to first have a try to fix this [...] at the source
That's a good plan, thanks for the pointer!

> I think there are non controversial parts in this PR that would have no issue to go upstream
I'll see about re-working this PR to avoid the PLL2 issues.

attie-argentum commented:
@erwango - please see the amended PR...

This adds the fundamental support for the STM32F105xx parts.

I see that CAN2 is something that has received some attention since I initially opened the PR, so I've added the device tree nodes for it.

I'll work on PLL2 separately with the STM32CubeF1 team, and when that's ready I'll look at opening a separate PR to integrate that here.

attie-argentum commented:
@erwango - any progress on this PR?

erwango commented on dts/arm/st/f1/stm32f105Xc.dtsi:
```diff
@@ -18,5 +18,24 @@
+		compatible = "usb-nop-xceiv";
+		#phy-cells = <0>;
+		label = "OTGFS_PHY";
 	};
```
This should be moved to stm32f105.dtsi so it could be available to stm32f107.dtsi

erwango commented on dts/arm/st/f1/stm32f105Xc.dtsi:
```diff
@@ -31,6 +31,40 @@
 			label = "OTGFS";
 		};
+
+		can1: can@40006400 {
+			compatible = "st,stm32-can";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40006400 0x400>;
+			interrupts = <19 0>, <20 0>, <21 0>, <22 0>;
+			interrupt-names = "TX", "RX0", "RX1", "SCE";
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x02000000>;
+			status = "disabled";
+			label = "CAN_1";
+			bus-speed = <250000>;
+			sjw = <1>;
+			prop-seg = <0>;
+			phase-seg1 = <5>;
+			phase-seg2 = <6>;
+		};
+
+		can2: can@40006800 {
+			compatible = "st,stm32-can";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40006800 0x400>;
+			interrupts = <63 0>, <64 0>, <65 0>, <66 0>;
+			interrupt-names = "TX", "RX0", "RX1", "SCE";
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x04000000>;
+			status = "disabled";
+			label = "CAN_2";
+			bus-speed = <250000>;
+			sjw = <1>;
+			prop-seg = <0>;
+			phase-seg1 = <5>;
+			phase-seg2 = <6>;
+		};
```
This should be moved to stm32f105.dtsi so it could be available to stm32f107.dtsi

erwango reviewed (changes requested):
2 comments otherwise LGTM

attie-argentum commented:
@erwango thanks! I've made these changes and rebased onto master.

erwango commented:
@ydamigos, @jfischer-no would you mind have a new look ?

ABOSTM commented on dts/arm/st/f1/stm32f105.dtsi:
```diff
@@ -85,5 +85,24 @@
+			interrupt-names = "otgfs";
+			num-bidir-endpoints = <4>;
+			ram-size = <1280>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x00001000>;
```
According to Reference Manual RM0008,
USB clock is on APB1 bit 23
```suggestion
			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00800000>;
```

attie-argentum replied:
`USB-FS` / USB OTG is in `RCC_AHBENR` bit 12 (which I believe makes the original correct).

Bit 23 of `RCC_APB1ENR` is reserved for the STM32F105... note it's a connectivity line device (thus Chapter 8, not Chapter 7, and Chapter 28 / `OTG_FS`, not Chapter 23 / `USB`).

ABOSTM replied:
Ok, your are right, my bad.

ABOSTM reviewed (approved):
LGTM
